pkg/dhclient/dhclient.go

```diff
@@ -0,0 +1,338 @@
+}
+
+func getClient4(iface netlink.Link, timeout time.Duration) (*dhcp4client.Client, error) {
+	mac := iface.Attrs().HardwareAddr
+	conn, err := dhcp4client.NewPacketSock(iface.Attrs().Index)
+	if err != nil {
+		return nil, fmt.Errorf("client connection generation: %v", err)
+	}
+
+	client, err := dhcp4client.New(dhcp4client.HardwareAddr(mac), dhcp4client.Connection(conn), dhcp4client.Timeout(timeout))
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func getClient6(iface netlink.Link, timeout time.Duration, retry int) (*dhcp6.Client, error) {
+	conn, err := dhcp6.NewPacketSock(iface.Attrs().Index)
+	if err != nil {
+		return nil, fmt.Errorf("client connection generation: %v", err)
+	}
+	return dhcp6.New(iface.Attrs().HardwareAddr, conn, timeout, retry), nil
+}
+
+type Client struct {
+	iface    netlink.Link
+	d4client *dhcp4client.Client
+	d6client *dhcp6.Client
+
+	retry int
+}
+
+type Version int
+
+const (
+	V4 Version = iota
+	V6
+)
+
+func New(version Version, iface netlink.Link, timeout time.Duration, retry int) (*Client, error) {
+	switch version {
+	case V4:
+		client, err := getClient4(iface, timeout)
+		if err != nil {
+			return nil, err
+		}
+		return &Client{
+			iface:    iface,
+			d4client: client,
+			retry:    retry,
+		}, nil
+
+	case V6:
+		client, err := getClient6(iface, timeout, retry)
+		if err != nil {
+			return nil, err
+		}
+		return &Client{
+			iface:    iface,
+			d6client: client,
+			retry:    retry,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("invalid DHCP version: %d", version)
+	}
+}
+
+func (c *Client) Solicit() (Packet, error) {
+	if c.d4client != nil {
+		return requestPacket4(c.d4client, c.retry)
+	} else if c.d6client != nil {
+		return solicitPacket6(c.d6client)
+	}
+	return nil, fmt.Errorf("invalid DHCP version?")
+}
+
+func (c *Client) Renew(p Packet) (Packet, error) {
+	if c.d4client != nil {
+		return renewPacket4(c.d4client, p, c.retry)
+	} else if c.d6client != nil {
+		return solicitPacket6(c.d6client)
+	}
+	return nil, fmt.Errorf("invalid DHCP version?")
+}
+
+func solicitPacket6(client *dhcp6.Client) (Packet, error) {
+	_, packet, err := client.Solicit()
+	if err != nil {
+		return nil, err
+	}
+	return newPacket6(packet)
+}
+
+// RequestPacket4 uses the given dhcp4client to solicit a DHCPv4 packet.
+func requestPacket4(client *dhcp4client.Client, retry int) (Packet, error) {
+	return getPacket4(func() (bool, dhcp4.Packet, error) {
+		return client.Request()
+	}, retry)
+}
+
+func renewPacket4(client *dhcp4client.Client, packet Packet, retry int) (Packet, error) {
+	pkg, ok := packet.(*packet4)
+	if !ok {
+		return nil, fmt.Errorf("passed non-DHCPv4 packet to RenewPacket4")
+	}
+
+	return getPacket4(func() (bool, dhcp4.Packet, error) {
+		return client.Renew(pkg.p)
+	}, retry)
+}
+
+func getPacket4(f func() (bool, dhcp4.Packet, error), retry int) (Packet, error) {
+	for i := 0; i < retry || retry < 0; i++ {
+		success, packet, err := f()
```

Comment:
here's a perfect place to stash statistics in a client. But we would need to either return said statistics or pass the client in.